File: packages/sdk/src/crdt/gcounter.ts
import { encode } from "../codec.js";
import type { WsTransport } from "../transport/websocket.js";
import type { GCounterDelta } from "../sync/delta.js";

export interface GCounterState {
  counts: Record<string, number>;
}

/**
 * Low-level handle for a grow-only counter (GCounter) CRDT.
 *
 * Obtained via `MeridianClient.gcounter()`. Prefer the `useGCounter` React hook
 * for component-level usage; use this handle directly in non-React environments.
 */
export class GCounterHandle {
  private state: GCounterState;
  private readonly clientId: number;
  private readonly crdtId: string;
  private readonly ns: string;
  private readonly transport: WsTransport;

  private readonly listeners = new Set<(value: number) => void>();

  constructor(opts: {
    ns: string;
    crdtId: string;
    clientId: number;
    transport: WsTransport;
    initial?: GCounterState;
  }) {
    this.ns = opts.ns;
    this.crdtId = opts.crdtId;
    this.clientId = opts.clientId;
    this.transport = opts.transport;
    this.state = opts.initial ?? { counts: {} };
  }

  /** Returns the current counter value (sum of all client contributions). */
  value(): number {
    return Object.values(this.state.counts).reduce((a, b) => a + b, 0);
  }

  /** Returns the raw per-client contribution map, keyed by client id string. */
  counts(): Readonly<Record<string, number>> {
    return this.state.counts;
  }

  /**
   * Registers a listener that is called whenever the counter value changes.
   *
   * @returns An unsubscribe function — call it to stop receiving updates.
   */
  onChange(listener: (value: number) => void): () => void {
    this.listeners.add(listener);
    return () => { this.listeners.delete(listener); };
  }

  /**
   * Increments the counter by `amount` (default `1`) and broadcasts the delta.
   *
   * @throws {RangeError} If `amount` is not greater than zero.
   */
  increment(amount: number = 1, ttlMs?: number): void {
    if (amount <= 0) throw new RangeError("GCounter: increment amount must be > 0");

    const key = String(this.clientId);
    this.state.counts[key] = (this.state.counts[key] ?? 0) + amount;
    this.emit();

    const op = encode({
      GCounter: {
        client_id: this.clientId,
        amount,
      },
    });
    this.transport.send({
      Op: { crdt_id: this.crdtId, op_bytes: op, ...(ttlMs !== undefined && { ttl_ms: ttlMs }) },
    });
  }

  applyDelta(delta: GCounterDelta): void {
    let changed = false;
    for (const [id, count] of Object.entries(delta.counters)) {
      const current = this.state.counts[id] ?? 0;
      if (count > current) {
        this.state.counts[id] = count;
        changed = true;
      }
    }
    if (changed) this.emit();
  }

  private emit(): void {
    const v = this.value();
    for (const listener of this.listeners) {
      listener(v);
    }
  }
}

File: packages/python/src/meridian/crdt/_gcounter.py
import asyncio
from collections.abc import AsyncIterator
from typing import TYPE_CHECKING

from .._codec import encode

if TYPE_CHECKING:
    from .._transport import Transport


class GCounter:
    """Grow-only counter CRDT. Increment-only, convergent across all nodes."""

    def __init__(self, crdt_id: str, client_id: int, transport: "Transport") -> None:
        self._id = crdt_id
        self._client_id = client_id
        self._transport = transport
        self._counts: dict[str, int] = {}
        self._listeners: list[asyncio.Queue[int]] = []


    def value(self) -> int:
        return sum(self._counts.values())

    def counts(self) -> dict[str, int]:
        return dict(self._counts)


    def increment(self, amount: int = 1, *, ttl_ms: int | None = None) -> None:
        if amount <= 0:
            raise ValueError("amount must be > 0")
        key = str(self._client_id)
        self._counts[key] = self._counts.get(key, 0) + amount
        self._notify()
        op = encode({"GCounter": {"client_id": self._client_id, "amount": amount}})
        msg: dict = {"Op": {"crdt_id": self._id, "op_bytes": op}}
        if ttl_ms is not None:
            msg["Op"]["ttl_ms"] = ttl_ms
        self._transport.send(msg)


    async def changes(self) -> AsyncIterator[int]:
        """Yields the counter value every time it changes."""
        q: asyncio.Queue[int] = asyncio.Queue()
        self._listeners.append(q)
        try:
            while True:
                yield await q.get()
        finally:
            self._listeners.remove(q)


    def _apply_delta(self, delta: dict) -> None:
        changed = False
        for k, v in (delta.get("counters") or {}).items():
            if v > self._counts.get(k, 0):
                self._counts[k] = v
                changed = True
        if changed:
            self._notify()

    def _notify(self) -> None:
        v = self.value()
        for q in self._listeners:
            try:
                q.put_nowait(v)
            except asyncio.QueueFull:
                pass
